File: modules/agent-go/internal/serializers/organization.go
package serializers

import (
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
)

// OrganizationListResponse represents an organization in list view
type OrganizationListResponse struct {
	ID                 int64     `json:"id"`
	Name               string    `json:"name"`
	Website            *string   `json:"website"`
	Description        *string   `json:"description"`
	EmployeeCountRange *string   `json:"employee_count_range"`
	FundingStage       *string   `json:"funding_stage"`
	Industries         []string  `json:"industries"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

// OrganizationDetailResponse represents an organization in detail view
type OrganizationDetailResponse struct {
	OrganizationListResponse
	LinkedInURL   *string           `json:"linkedin_url"`
	CrunchbaseURL *string           `json:"crunchbase_url"`
	Phone         *string           `json:"phone"`
	Technologies  []TechnologyBrief `json:"technologies"`
	FundingRounds []FundingBrief    `json:"funding_rounds"`
}


// TechnologyBrief represents technology summary
type TechnologyBrief struct {
	ID       int64   `json:"id"`
	Name     string  `json:"name"`
	Category *string `json:"category"`
}

// ContactBrief represents contact summary
type ContactBrief struct {
	ID        int64   `json:"id"`
	FirstName string  `json:"first_name"`
	LastName  string  `json:"last_name"`
	Email     *string `json:"email"`
	Phone     *string `json:"phone"`
	Title     *string `json:"title"`
	IsPrimary bool    `json:"is_primary"`
}

// FundingBrief represents funding round summary
type FundingBrief struct {
	ID          int64      `json:"id"`
	RoundType   string     `json:"round_type"`
	Amount      *int64     `json:"amount"`
	AnnouncedAt *time.Time `json:"announced_at"`
}

// OrganizationToListResponse converts Organization model to list response
func OrganizationToListResponse(org *models.Organization) OrganizationListResponse {
	resp := OrganizationListResponse{
		ID:          org.ID,
		Name:        org.Name,
		Website:     org.Website,
		Description: org.Description,
		CreatedAt:   org.CreatedAt,
		UpdatedAt:   org.UpdatedAt,
	}

	if org.EmployeeCountRange != nil {
		resp.EmployeeCountRange = &org.EmployeeCountRange.Label
	}

	if org.FundingStage != nil {
		resp.FundingStage = &org.FundingStage.Label
	}

	// Get industries from Account relation
	if org.Account != nil && org.Account.Industries != nil {
		resp.Industries = make([]string, len(org.Account.Industries))
		for i, ind := range org.Account.Industries {
			resp.Industries[i] = ind.Name
		}
	}

	return resp
}

// OrganizationToDetailResponse converts Organization model to detail response
func OrganizationToDetailResponse(org *models.Organization) OrganizationDetailResponse {
	resp := OrganizationDetailResponse{
		OrganizationListResponse: OrganizationToListResponse(org),
		LinkedInURL:              org.LinkedInURL,
		CrunchbaseURL:            org.CrunchbaseURL,
		Phone:                    org.Phone,
	}

	if org.FundingRounds != nil {
		resp.FundingRounds = make([]FundingBrief, len(org.FundingRounds))
		for i, f := range org.FundingRounds {
			resp.FundingRounds[i] = FundingBrief{
				ID:          f.ID,
				RoundType:   f.RoundType,
				Amount:      f.Amount,
				AnnouncedAt: f.AnnouncedDate,
			}
		}
	}

	return resp
}

// OrgTechnologyResponse represents a technology linked to an organization
type OrgTechnologyResponse struct {
	OrganizationID int64    `json:"organization_id"`
	TechnologyID   int64    `json:"technology_id"`
	Source         *string  `json:"source"`
	Confidence     *float64 `json:"confidence"`
	DetectedAt     string   `json:"detected_at"`
}

// FundingRoundResponse represents a funding round in API responses
type FundingRoundResponse struct {
	ID             int64   `json:"id"`
	OrganizationID int64   `json:"organization_id"`
	RoundType      string  `json:"round_type"`
	Amount         *int64  `json:"amount"`
	AnnouncedDate  *string `json:"announced_date"`
	LeadInvestor   *string `json:"lead_investor"`
	SourceURL      *string `json:"source_url"`
}

File: modules/agent/src/serializers/organization.py
"""Serializers for organization-related models."""

from serializers.common import contact_to_dict
from serializers.account import get_account_relationships


def technology_to_dict(org_tech) -> dict:
    """Convert organization technology ORM to response dictionary."""
    return {
        "organization_id": org_tech.organization_id,
        "technology_id": org_tech.technology_id,
        "technology": {
            "id": org_tech.technology.id,
            "name": org_tech.technology.name,
            "category": org_tech.technology.category,
            "vendor": org_tech.technology.vendor,
        } if org_tech.technology else None,
        "detected_at": org_tech.detected_at.isoformat() if org_tech.detected_at else None,
        "source": org_tech.source,
        "confidence": float(org_tech.confidence) if org_tech.confidence else None,
        "created_at": org_tech.created_at.isoformat() if org_tech.created_at else None,
        "updated_at": org_tech.updated_at.isoformat() if org_tech.updated_at else None,
    }


def funding_round_to_dict(fr) -> dict:
    """Convert funding round ORM to response dictionary."""
    return {
        "id": fr.id,
        "organization_id": fr.organization_id,
        "round_type": fr.round_type,
        "amount": fr.amount,
        "announced_date": fr.announced_date.isoformat() if fr.announced_date else None,
        "lead_investor": fr.lead_investor,
        "source_url": fr.source_url,
        "created_at": fr.created_at.isoformat() if fr.created_at else None,
        "updated_at": fr.updated_at.isoformat() if fr.updated_at else None,
    }


def signal_to_dict(s) -> dict:
    """Convert signal ORM to response dictionary."""
    return {
        "id": s.id,
        "account_id": s.account_id,
        "signal_type": s.signal_type,
        "headline": s.headline,
        "description": s.description,
        "signal_date": s.signal_date.isoformat() if s.signal_date else None,
        "source": s.source,
        "source_url": s.source_url,
        "sentiment": s.sentiment,
        "relevance_score": float(s.relevance_score) if s.relevance_score else None,
        "created_at": s.created_at.isoformat() if s.created_at else None,
        "updated_at": s.updated_at.isoformat() if s.updated_at else None,
    }


def org_to_list_response(org) -> dict:
    """Convert organization ORM to response dictionary - optimized for list/table view."""
    industries = []
    if org.account and org.account.industries:
        industries = [ind.name for ind in org.account.industries]

    return {
        "id": org.id,
        "name": org.name,
        "website": org.website,
        "industries": industries,
        "employee_count_range": org.employee_count_range.label if org.employee_count_range else None,
        "funding_stage": org.funding_stage.label if org.funding_stage else None,
        "description": org.description,
    }


def org_to_search_response(org) -> dict:
    """Convert organization ORM to response dictionary - minimal for search/autocomplete."""
    industries = []
    if org.account and org.account.industries:
        industries = [ind.name for ind in org.account.industries]

    return {
        "id": org.id,
        "name": org.name,
        "industries": industries,
    }


def org_to_detail_response(org, contacts=None, include_research=False) -> dict:
    """Convert organization ORM to full response dictionary - for detail/edit views."""
    industries = []
    if org.account and org.account.industries:
        industries = [ind.name for ind in org.account.industries]

    projects = []
    if org.account and org.account.projects:
        projects = [{"id": p.id, "name": p.name} for p in org.account.projects]

    result = {
        "id": org.id,
        "account_id": org.account_id,
        "name": org.name,
        "website": org.website,
        "phone": org.phone,
        "industries": industries,
        "projects": projects,
        "employee_count": org.employee_count,
        "employee_count_range_id": org.employee_count_range_id,
        "employee_count_range": org.employee_count_range.label if org.employee_count_range else None,
        "funding_stage_id": org.funding_stage_id,
        "funding_stage": org.funding_stage.label if org.funding_stage else None,
        "description": org.description,
        "revenue_range_id": org.revenue_range_id,
        "revenue_range": org.revenue_range.label if org.revenue_range else None,
        "founding_year": org.founding_year,
        "headquarters_city": org.headquarters_city,
        "headquarters_state": org.headquarters_state,
        "headquarters_country": org.headquarters_country,
        "company_type": org.company_type,
        "linkedin_url": org.linkedin_url,
        "crunchbase_url": org.crunchbase_url,
        "created_at": org.created_at.isoformat() if org.created_at else None,
        "updated_at": org.updated_at.isoformat() if org.updated_at else None,
    }

    if contacts is not None:
        result["contacts"] = [contact_to_dict(c) for c in contacts]

    # Get relationships from Account level
    if org.account:
        result["relationships"] = get_account_relationships(org.account, org.account_id)
    else:
        result["relationships"] = []

    if include_research:
        result["technologies"] = [technology_to_dict(t) for t in (org.technologies or [])]
        result["funding_rounds"] = [funding_round_to_dict(fr) for fr in (org.funding_rounds or [])]
        # Signals are now on the Account level
        account_signals = org.account.signals if org.account else []
        result["signals"] = [signal_to_dict(s) for s in (account_signals or [])]

    return result
